openverse_catalog/dags/providers/provider_api_scripts/finnish_museums.py

```diff
@@ -1,167 +1,128 @@
-    total_images = 0
-    license_url = get_license_url(obj)
-    if license_url is None:
-        return None
-    foreign_identifier = obj.get("id")
-    title = obj.get("title")
-    building = obj.get("buildings")[0].get("value")
-    source = next((s for s in sub_providers if building in sub_providers[s]), provider)
-    foreign_landing_url = _get_landing(obj)
-    raw_tags = _get_raw_tags(obj)
-    image_list = obj.get("images")
-    for img in image_list:
-        image_url = _get_image_url(img)
-        total_images = image_store.add_item(
-            license_info=get_license_info(license_url=license_url),
-            foreign_identifier=foreign_identifier,
-            foreign_landing_url=foreign_landing_url,
-            image_url=image_url,
-            title=title,
-            source=source,
-            raw_tags=raw_tags,
+class FinnishMuseumsDataIngester(ProviderDataIngester):
+    providers = {"image": PROVIDER}
+    endpoint = "https://api.finna.fi/api/v1/search"
+    batch_limit = 100
+    delay = 5
+    format_type = "0/Image/"
+
+    def ingest_records(self, **kwargs):
+        for building in BUILDINGS:
+            logger.info(f"Obtaining images of building {building}")
+            super().ingest_records(building=building)
+
+    def get_next_query_params(self, prev_query_params, **kwargs):
+        if not prev_query_params:
+            building = kwargs.get("building")
+            return {
+                "filter[]": [f'format:"{self.format_type}"', f'building:"{building}"'],
+                "limit": self.batch_limit,
+                "page": 1,
+            }
+        return {**prev_query_params, "page": prev_query_params["page"] + 1}
+
+    def get_media_type(self, record):
+        return "image"
+
+    def get_batch_data(self, response_json):
+        if (
+            response_json is None
+            or str(response_json.get("status")).lower() != "ok"
+            or response_json.get("records") is None
+            or len(response_json.get("records")) == 0
+        ):
+            return None
+
+        return response_json.get("records")
+
+    def get_record_data(self, data):
+        records = []
+
+        license_url = self.get_license_url(data)
+        if license_url is None:
+            return None
+
+        foreign_identifier = data.get("id")
+        title = data.get("title")
+        building = data.get("buildings")[0].get("value")
+        source = next(
+            (s for s in SUB_PROVIDERS if building in SUB_PROVIDERS[s]), PROVIDER
         )
-    return total_images
+        foreign_landing_url = self._get_landing(data)
+        raw_tags = self._get_raw_tags(data)
+
+        image_list = data.get("images")
+        for img in image_list:
+            image_url = self._get_image_url(img)
+            records.append(
+                {
+                    "license_info": get_license_info(license_url),
+                    "foreign_identifier": foreign_identifier,
```

Comment:
Foreign identifier is supposed to be unique. Here, we can be collecting several images for a single record. They will all have the same foreign_identifier, resulting in many duplicates. We need to either add the image index to the `foreign_identifier` here, or only use the first image from the `image_list` here that would be the single image for the record in question.

I tried looking into the records that have several images, and I think I'm in favor of using only the first image for each record.
Here's an example of record with 2 images that are quite different and we _might_ want to add both: https://www.finna.fi/Record/museovirasto.AA6BD6DA6AB5F4C526C79DC74C7A58D7?imgid=1

But there are many more records that are like this: https://www.finna.fi/Record/museovirasto.D24CDFCAA0324B50C830BD43133CA3F1
The first image is the actual image, and all others are pictures of the back of the picture frame.

Reply:
Yep, we ran into this awhile ago, including the funny back-of-frame issue 😭 -- there's a separate ticket for it here https://github.com/WordPress/openverse-catalog/issues/450 but I think your assessment is exactly right. Since this involves making a decision about how to handle the duplicates I'd still prefer to keep that discussion/work separate.